processing/src/main/java/org/apache/druid/query/operator/Operator.java

```diff
@@ -22,72 +22,48 @@
 {
   /**
-   * Called to initiate the lifecycle of the Operator.  If an operator needs to checkout resources or anything to do
-   * its work, this is probably the place to do it.
+   * Tells the Operation to start doing its work.  Data will be pushed into the Receiver.
    *
-   * Work should *never* be done in this method, this method only exists to acquire resources that are known to be
-   * needed before doing any work.  As a litmus test, if there is ever a call to `op.next()` inside of this method,
-   * then something has been done wrong as that call to `.next()` is actually doing work.  Such code should be moved
-   * into being lazily evaluated as part of a call to `.next()`.
+   * @param receiver a receiver that will receive data
    */
-  void open();
+  void go(Receiver receiver);
 
-  /**
-   * Returns the next RowsAndColumns object that the Operator can produce.  Behavior is undefined if
-   * {@link #hasNext} returns false.
-   *
-   * @return the next RowsAndColumns object that the operator can produce
-   */
-  RowsAndColumns next();
+  interface Receiver
+  {
+    /**
+     * Used to push data.  Return value indicates if more data will be accepted.  If false, push should not
+     * be called anymore.
+     *
+     * @param rac {@link RowsAndColumns} of data
+     * @return a boolean value indicating if more data will be accepted.  If false, push should never be called
+     * anymore
+     */
+    boolean push(RowsAndColumns rac);
 
-  /**
-   * Used to identify if it is safe to call {@link #next}
-   *
-   * @return true if it is safe to call {@link #next}
-   */
-  boolean hasNext();
-
-  /**
-   * Closes this Operator.  The cascade flag can be used to identify that the intent is to close this operator
-   * and only this operator without actually closing child operators.  Other databases us this sort of functionality
-   * with a planner that is watching over all of the objects and force-closes even if they were closed during normal
-   * operations.  In Druid, in the data pipeline where this was introduced, we are guaranteed to always have close
-   * called regardless of errors or exceptions during processing, as such, at time of introduction, there is no
-   * call that passes false for cascade.
-   * <p>
-   * That said, given that this is a common thing for these interfaces for other databases, we want to preserve the
-   * optionality of being able to leverage what they do.  As such, we define the method this way with the belief
-   * that it might be used in the future.  Semantically, this means that all implementations of Operators must
-   * expect to be closed multiple times.  I.e. after being closed, it is an error for open, next or hasNext to be
-   * called, but close can be called any number of times.
-   *
-   * @param cascade whether to call close on child operators.
-   */
-  void close(boolean cascade);
+    /**
+     * Used to indicate that no more data will ever come
+     */
+    void completed();
```

Comment:
is there any contract around what happens if this is called more than once? what happens if i call `push`? is this `close` with a different name? (not to imply i'm advocating for this to be a `Closeable`, i'm really not)

Reply:
Good question, this is definitely not `close()` with a different name and I should call that out better in the javadoc.  It is more correct to say that this is `flush()` with a different name, except it's expected that after `flush()` no more `write()`s would happen.  It's probably going to be tempting to try to close resources in here, but there is no guarantee that `complete()` will be called if an exception destroys the operations during a `push()` call.